runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10283,45 +10269,111 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
+               {
+               shiftReg = cg->allocateRegister();
+               if (conditions != NULL)
+                  conditions->addPostCondition(shiftReg, TR::RealRegister::AssignAny);
+
+               iCursor = generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, addressReg, resReg, iCursor);
+               iCursor = generateLoad32BitConstant(cg, node, loops, shiftReg, true, iCursor, conditions);
+               iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerLoopLabel);
+               iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 255, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+               iCursor = generateRXInstruction(cg, TR::InstOpCode::LA, node, addressReg, generateS390MemoryReference(addressReg, 256, cg), iCursor);
+               iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, shiftReg, fillerLoopLabel);
+               loops = 0;
+               }
 
-         // minus 1 from lengthreg since xc auto adds 1 to it
+            while (loops > 0)
+               {
+               iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 255, generateS390MemoryReference(addressReg, displacement, cg), generateS390MemoryReference(addressReg, displacement, cg), iCursor);
+               displacement += 256;
+               loops--;
+               }
 
-         iCursor = generateEXDispatch(node, cg, lengthReg, shiftReg, iCursor);
-         iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, doneLabel);
+            switch (remainder)
+               {
+               case 0:
+                  break;
+               case 1:
+                  iCursor = generateSIInstruction(cg, TR::InstOpCode::MVI, node, generateS390MemoryReference(addressReg, displacement, cg), 0, iCursor);
+                  break;
+               case 2:
+                  iCursor = generateSILInstruction(cg, TR::InstOpCode::MVHHI, node, generateS390MemoryReference(addressReg, displacement, cg), 0, iCursor);
+                  break;
+               case 4:
+                  iCursor = generateSILInstruction(cg, TR::InstOpCode::MVHI, node, generateS390MemoryReference(addressReg, displacement, cg), 0, iCursor);
+                  break;
+               case 8:
+                  iCursor = generateSILInstruction(cg, TR::InstOpCode::MVGHI, node, generateS390MemoryReference(addressReg, displacement, cg), 0, iCursor);
+                  break;
+               default:
+                  iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, (remainder - 1), generateS390MemoryReference(addressReg, displacement, cg),
+                                                   generateS390MemoryReference(addressReg, displacement, cg), iCursor);
+                  break;
+               }
+            }
+            cg->stopUsingRegister(addressReg);
+            cg->stopUsingRegister(shiftReg);
```

Comment:
We should call stopUsingRegister on shiftReg if it was allocated (Right now you will run into issue where you will have NULL dereference.

Reply:
`stopUsingRegister` checks for null registers